Support for LargeUtf8 and LargeBinary

chitralverma opened the pull request:
# Description
Added missing mapping from below mentioned arrow types to delta types,
- `LargeUtf8` (LargeString) -> `string`
- `LargeBinary` -> `binary`

# Related Issue(s)
closes https://github.com/delta-io/delta-rs/issues/1024

chitralverma commented:
@houqp  this should be enough to close https://github.com/delta-io/delta-rs/issues/1024  right?

wjones127 commented:
@chitralverma I would recommend writing a unit test for the Python writer to be sure. IIRC we validate the schema, so we might need a modified equality function that ignores the large string difference.

houqp commented:
+1 to what @wjones127 said :) Thanks @chitralverma for taking a stab at this.

chitralverma commented:
cool, I'll add some tests on python side

BMeyn commented:
@chitralverma Any updates about the python tests? I'm happy to support when you are to busy.

ritchie46 commented:
@chitralverma any news? This would enable a delta-writer in polars as well. :+1:

chitralverma commented:
I will work on the tests that are pending for this and rebase the changes soon

alexwilcoxson-rel commented:
Should `FixedSizedBinary` from arrow be mapped to "binary" as well?

chitralverma commented:
> Should `FixedSizedBinary` from arrow be mapped to "binary" as well?

added

wjones127 commented:
@chitralverma The schema isn't derived from the Parquet file but should be derived from the table schema. Delta Lake has [its own standard](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#schema-serialization-format) for describing table schemas and its doesn't differentiate these two types. If it helps, you could write a function for converting delta schemas to arrow schemas that takes an option to use large string and large binary.

This is the line where that converts back:

https://github.com/delta-io/delta-rs/blob/70a9d095bfa7bcffaecaff3dc9404282366848b2/rust/src/delta_arrow.rs#L91

chitralverma commented:
@wjones127 made the changes as per your suggestions.

Added an option of `as_large_types` which converts the fields of Delta Schema to pyarrow schema with large types.

chitralverma commented:
@wjones127 I have fixed the linting issues, and updated with the main branch. I also added support for uint* and float16 as well. Now only the following types remain, any ideas for these because I don't think we can map all of them?

            DataType::Null => {}
            DataType::Date64 => {}
            DataType::Time32(_) => {}
            DataType::Time64(_) => {}
            DataType::Duration(_) => {}
            DataType::Interval(_) => {}
            DataType::Union(_, _, _) => {}
            DataType::Dictionary(_, _) => {}
            DataType::RunEndEncoded(_, _) => {}

chitralverma commented:
> @wjones127 I have fixed the linting issues, and updated with the main branch. I also added support for uint* and float16 as well. Now only the following types remain, any ideas for these because I don't think we can map all of them?
> 
> ```
>         DataType::Null => {}
>         DataType::Time32(_) => {}
>         DataType::Time64(_) => {}
>         DataType::Duration(_) => {}
>         DataType::Interval(_) => {}
>         DataType::Union(_, _, _) => {}
>         DataType::Dictionary(_, _) => {}
>         DataType::RunEndEncoded(_, _) => {}
> ```

From the above remaining types, here is a proposed strategy,

- Time32 => Int32
- Time64 => Int64
- Duration => Int64
- Interval => Int64
- Union | Dictionary | RunEndEncoded => Error as Not Supported
- Null => String or Error as Not Supported

Any suggestions on this ?

stinodego commented:
Personally, I'd suggest "Error as not supported" for all of those. That requires the user to explicitly acknowledge that they are writing unsupported types, and cast to supported types themselves.

wjones127 commented on python/src/schema.rs:
```diff
@@ -1008,14 +1008,69 @@ impl PySchema {
+                    field.with_data_type(ArrowDataType::LargeBinary)
+                }
+
+                ArrowDataType::List(f) | ArrowDataType::FixedSizeList(f, _) => {
+                    let sub_field = convert_to_large_type(*f.clone(), f.data_type().clone());
+                    field.with_data_type(ArrowDataType::LargeList(Box::from(sub_field)))
+                }
```
Why convert fixed size list to large list? It doesn't have any indices, so it doesn't have a small / large variant. Is this just because Polars doesn't support FSL?

wjones127 replied:
It doesn't seem like this is necessary to make the PR work, so I'd propose we change this to keep FSB and FSL as the same type. We still should transform the inner type of FSL like we do other nested types.

chitralverma replied:
yes you are correct, this was mostly for compatibility, but we can also do the cast or error handling for unsupported types in Polars directly.

Made the requested changes.

wjones127 commented on python/src/schema.rs:
```diff
@@ -1008,14 +1008,69 @@ impl PySchema {
     /// :rtype: pyarrow.Schema
-    #[pyo3(text_signature = "($self)")]
-    fn to_pyarrow(self_: PyRef<'_, Self>) -> PyResult<PyArrowType<ArrowSchema>> {
+    #[pyo3(text_signature = "($self, as_large_types)")]
```
I think this should handle the default value:
```suggestion
    #[pyo3(signature = "(as_large_types=false)")]
```

wjones127 commented on python/src/schema.rs:
```diff
@@ -1008,14 +1008,69 @@ impl PySchema {
     /// Return equivalent PyArrow schema
     ///
```
```suggestion
    ///
    /// :param as_large_types: get schema with all variable size types (list,
    ///   binary, string) as large variants (with int64 indices). This is for
    ///   compatibility with systems like Polars that only support the large 
    ///   versions of Arrow types.
    ///
```

chitralverma replied:
Made the requested changes.

wjones127 commented on python/src/schema.rs:
```diff
@@ -1008,14 +1008,69 @@ impl PySchema {
-    #[pyo3(text_signature = "($self)")]
-    fn to_pyarrow(self_: PyRef<'_, Self>) -> PyResult<PyArrowType<ArrowSchema>> {
+    #[pyo3(text_signature = "($self, as_large_types)")]
+    fn to_pyarrow(
+        self_: PyRef<'_, Self>,
+        as_large_types: Option<bool>,
```
```suggestion
    fn to_pyarrow(
        self_: PyRef<'_, Self>,
        as_large_types: bool,
```

wjones127 reviewed (changes requested):
Thanks for sticking with this @chitralverma 

I have a few changes suggested. Once addressed I think this is good to merge.

chitralverma commented:
> Thanks for sticking with this @chitralverma
> 
> I have a few changes suggested. Once addressed I think this is good to merge.

@wjones127 Thanks for the review comments, I have made the requested changes.

chitralverma commented:
some tests timeout

wjones127 commented:
Failures are unrelated. GitHub is being rather weird today.

chitralverma commented:
Thanks @wjones127
